src/daemon/distribution.c

```diff
@@ -0,0 +1,148 @@
+#include <math.h>
+#include "collectd.h"
+#include "distribution.h"
+
+struct bucket_s{
+	size_t bucket_counter;
+	double min_boundary;
+	double max_boundary;
+};
+
+struct distribution_s {
+     bucket_t *buckets;
+     size_t num_buckets;
+     size_t total_scalar_count;
+     double raw_data_sum;
+};
+
+//bucket constructor
+bucket_t initialize_bucket(double min_boundary, double max_boundary) {
+    bucket_t new_bucket = {
+        .bucket_counter = 0,
+        .min_boundary = min_boundary,
+        .max_boundary = max_boundary,
+    };
+    return new_bucket;
+} 
+
+//increment bucket_counter of specific bucket
+void increment(bucket_t *bucket) { 
+    bucket->bucket_counter++;
+} 
+
+distribution_t * distribution_new_linear(size_t num_buckets, double size) {
+    distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+    new_distribution->buckets = calloc(num_buckets, sizeof(bucket_t));
+
+   for(size_t i = 0; i < num_buckets; i++)
+    {
+        if(i == 0)
+        {
+            new_distribution->buckets[i] = initialize_bucket(0, size);
+        }
+        else if(i < num_buckets - 1)
+        {
+            new_distribution->buckets[i] = initialize_bucket(i * size + 1, i*size + size);
+        }
+        else 
+        {
+            new_distribution->buckets[i] = initialize_bucket(new_distribution->buckets[i - 1].max_boundary + 1, INFINITY);
+        }
+    }
+    
+    new_distribution->num_buckets = num_buckets;
+    new_distribution->total_scalar_count = 0;
+    new_distribution->raw_data_sum = 0;
+}
+
+distribution_t * distribution_new_exponential(size_t num_buckets, double initial_size, double factor) {
+    distribution_t *new_distribution = calloc(1, sizeof(distribution_t));
+    new_distribution->buckets = calloc(num_buckets, sizeof(bucket_t));
+    double previous_bucket_size;
+    double new_max;
+
+    for(int i = 0; i < num_buckets; i++)
+    {
+        if(i == 0)
+        {
+            new_distribution->buckets[i] = initialize_bucket(0, initial_size);
+        }
+        else if(i < num_buckets - 1)
+        {
+            previous_bucket_size = new_distribution->buckets[i - 1].max_boundary - new_distribution->buckets[i - 1].min_boundary;
+            new_max = new_distribution->buckets[i - 1].max_boundary + previous_bucket_size * factor;
+            new_distribution->buckets[i] = initialize_bucket(new_distribution->buckets[i - 1].max_boundary + 1, new_max);
+        }
+        else
+        {
+            new_distribution->buckets[i] = initialize_bucket(new_distribution->buckets[i - 1].max_boundary + 1, INFINITY);
+        }
+        
+    }
+    
+    new_distribution->num_buckets = num_buckets;
+    new_distribution->total_scalar_count = 0;
+    new_distribution->raw_data_sum = 0;
+}
```

Comment:
This function is missing a `return` statement:

```c
return new_distribution;
```

Reply:
Thank you, I was thinking in terms of Java constructors and forgot I was in C. I fixed it now.